lib/msal-node/src/cache/CacheManager.ts

```diff
@@ -0,0 +1,94 @@
+import { Storage } from './Storage';
+import { Serializer, Deserializer, InMemoryCache, JsonCache } from '@azure/msal-common';
+import { ICachePlugin } from './ICachePlugin';
+
+const defaultSerializedCache: JsonCache = {
+    Account: {},
+    IdToken: {},
+    AccessToken: {},
+    RefreshToken: {},
+    AppMetadata: {}
+};
+
+
+export class CacheManager {
+
+    hasChanged: boolean;
+    storage: Storage;
+    persistence: ICachePlugin;
+
+    constructor(storage: Storage, cachePlugin?: ICachePlugin) {
+        this.hasChanged = false;
+        this.storage = storage;
+        this.storage.registerChangeEmitter(this.handleChangeEvent.bind(this))
+        if (cachePlugin) {
+            this.setPersistence(cachePlugin);
+        }
+
+    }
+
+    setPersistence(persistence: ICachePlugin): void {
+        this.persistence = persistence;
+    }
+
+    handleChangeEvent() {
+        this.hasChanged = true;
+    }
+
+    mergeState(oldState: JsonCache, currentState: JsonCache) {
+        // TODO
+        // mergeUpdates(old, new)
+        // mergeRemovals(old, new)
+        return {
+            ...oldState,
+            ...currentState
+        };
+    }
+
+    //TODO think about separating serialize / deserialize from writeToPersistance and readFromPersistance
+
+    async serialize(): Promise<string> {
+        const cache: string = JSON.stringify(Serializer.serializeAllCache(this.storage.getCache()));
+        if (this.persistence) {
+            await this.persistence.writeToStorage((stateFromDisk) => {
```

Comment:
I would argue not to put the callback inline, makes it difficult to read:

```js
if (this.persistence) {
  const getMergedState = (stateFromDisk) => {
    const jsonFromDisk = JSON.parse(stateFromDisk);
    const allCachedSerliazed = Serializer.serializeAllCache(this.storage.getCache());
    const mergedState = this.mergeState(jsonFromDisk, allCachedSerliazed);
    return JSON.stringify(mergedState, null, 4);
  }
  await this.persistence.writeToStorage(getMergedState);
}

Reply:
Also, combining callbacks and promises like this makes for a weird API, and I'm not sure I'm a fan of the ergonomics.

Reply:
fixed

Reply:
Not sure there is a better ergonomic solution here, and its an internal implementation a customer never really deals with